### PR TITLE
Refactoring of deinflection of verbs

### PR DESCRIPTION
**Pull request by Enellis**
_Related to #1811._

### Overview
This pull request introduces three new intermediate types for deinflection of Godan verbs:
_(Ichidan verbs, with their single stem, would not benefit from these changes.)_
- TaTeStem
- DaDeStem
- MasuStem

These intermediate types allow for the abstraction of inflections that follow them. Specifically:

- After TaTeStem/DaDeStem-stem:
  - て-form and past (た-)form 
  - たら
  - たり
  - とく
  - ちゃう
    
- After MasuStem:
  - All forms of the polite helper verb ます
  - なさい
  - そう
  - たい
  - すぎる
 
To help with that, `Reason.None` has been introduced and is used when forwarding TaTe/DaDe-stem forms to the plain verb form. This step is irrelevant for subsequent data processing, so while deinflection, it will never be added to the `reasons` array.

### Benefits
- The specialized implementations for all the aforementioned forms could be deleted, as they are now covered by the new intermediate types.
- Many irregular verbs exhibit irregularities just in their て-stem or ます-stem. With these changes, it becomes sufficient to implement the irregular stem, ensuring that all subsequent inflections are correctly parsed.

### CODE REVIEW

**Review by birtles — COMMENTED** [review]
In general, this looks great. Thank you!

Two comments:

1) Do you think we need to update the description of `DeinflectRule` given these recent changes?

    (While we're at it, perhaps we should (i) move `DeinflectRule` and `Type` before `deinflectRuleData`, (ii) Make `DeinflectRule` a `type` rather than an `interface`, (iii) label the tuple elements of `deinflectRuleData`, e.g. `[from: string, to: string, toType: number, fromType: number, reason: Reason]`?)
    
2) I wonder how important defining `InitialOrMasu` and `IchidanOrKuru` are? My hunch is that it would be simpler and clearer just inlining the definitions for these?

**Comment by Enellis** [comment]
Likewise, thank you for your input!

> Do you think we need to update the description of `DeinflectRule` given these recent changes?

You're right, I completely overlooked this. I will update the description to reflect these changes including adding a few lines about the new intermediate types.

> (While we're at it, perhaps we should (i) move `DeinflectRule` and `Type` before `deinflectRuleData`, (ii) Make `DeinflectRule` a `type` rather than an `interface`, (iii) label the tuple elements of `deinflectRuleData`, e.g. `[from: string, to: string, toType: number, fromType: number, reason: Reason]`?)
>
> 2. I wonder how important defining `InitialOrMasu` and `IchidanOrKuru` are? My hunch is that it would be simpler and clearer just inlining the definitions for these?

The main reason for defining `InitialOrMasu` and `IchidanOrKuru` separately was because inlining these definitions causes the lines to become too long, leading to wrapping which makes the `deinflectRuleData` array a lot harder to read for me. The same would apply to changing `DeinflectRule` to a type, I suppose.
Is there an option to tell Prettier not to wrap the contents of this specific array? Without the wrapping issue I'd be happy to make all of these changes and inline the definitions again.

**Comment by birtles** [comment]
> > Do you think we need to update the description of `DeinflectRule` given these recent changes?
> 
> You're right, I completely overlooked this. I will update the description to reflect these changes including adding a few lines about the new intermediate types.

Thank you!

> > 2. I wonder how important defining `InitialOrMasu` and `IchidanOrKuru` are? My hunch is that it would be simpler and clearer just inlining the definitions for these?
> 
> The main reason for defining `InitialOrMasu` and `IchidanOrKuru` separately was because inlining these definitions causes the lines to become too long, leading to wrapping which makes the `deinflectRuleData` array a lot harder to read for me. The same would apply to implementing (ii) and (iii), I suppose. Is there an option to tell Prettier not to wrap the contents of this specific array? Without the wrapping issue I'd be happy to make all of these changes and inline the definitions again.

Ah, yes, that makes sense.

Yes, you can tell Prettier not to wrap the array by preceding it with `// prettier-ignore`: https://prettier.io/docs/en/ignore.html#javascript

**Comment by Enellis** [comment]
Update:
- To be honest, I struggled a bit being precise and concise updating the `DeinflectRule` description. I hope it is clear enough now. Otherwise I'm very open for any suggestions.
- When changing `DeinflectRule` to a named tuple, we would lose the ability to access the elements by dot notation later, like in `getDeinflectRuleGroups()`. When changing DeinflectRule to an object the whole `deinflectRuleData` becomes very verbose. What do you think is the best approach? 

I will stash the commits after your review.

**Comment by Enellis** [comment]
Thinking about it, I’d like to change the "reason" field to an array to eliminate compound forms like PolitePast, PolitePastNegative, etc. This would be an user-facing change because for example, 食べませんでした would be parsed as `< polite < past < negative` (or better `< polite < negative < past` as this reflects the grammatical structure more closely) instead of `< polite past negative`.

In the future, this would allow us to parse forms like 食べへんかった as `< dialectal < negative < past` or 食べまして as `< polite < te`. 

Would you be okay with this change? And if so, should that be a separate PR?

File: src/background/deinflect.ts [hunk 1]
```diff
@@ -918,30 +87,34 @@ const enum Type {
...
+  // is the intermediate type masu-stem while the to type is a godan/u-verb.
+  //
+  // Intermediate types ensure that some rules are only applied in specific
+  // conditions.
   //
-  // The fromType may also have the special value Initial which means it accepts
-  // anything BUT one of the other word types.
+  // For example, consider the deinflection rule "ます"→"" without the to type
+  // masu-stem. Then everything following, not just the correct masu stem, would
+  // match it. "食べろます" would be parsed as "< imperative < polite", which
```
Comment (birtles) [inline src/background/deinflect.ts:113]:
```suggestion
  // For example, consider the deinflection rule "ます"→"". Without the to-type
  // masu-stem everything following--not just the correct masu stem--would
  // match it. "食べろます" would be parsed as "< imperative < polite", which
```

File: src/background/deinflect.ts [hunk 2]
```diff
@@ -918,30 +87,34 @@ const enum Type {
...
   to: string;
   // A bit mask representing the type of words to which this rule can be
   // applied.
   //
-  // For example, 遊びすぎる would match the びすぎる→ぶ rule where the from
-  // type is an ichidan/ru-verb while the to type is a godan/u-verb.
+  // For example, 遊びすぎる would match the "すぎる"→"" rule where the from type
+  // is an ichidan/ru-verb while the to type is the intermediate type masu-stem.
+  // The remaining 遊び would then match the "び"→"ぶ" rule where the from type
+  // is the intermediate type masu-stem while the to type is a godan/u-verb.
```
Comment (birtles) [inline src/background/deinflect.ts:106]:
```suggestion
  // For example, 遊びすぎる would match the "すぎる"→"" rule where the from-type
  // is an ichidan/ru-verb while the to-type is the intermediate type masu-stem.
  // The remaining 遊び would then match the "び"→"ぶ" rule where the from-type
  // is the intermediate type masu-stem while the to-type is a godan/u-verb.
```

File: src/background/deinflect.ts [hunk 3]
```diff
@@ -918,30 +87,34 @@ const enum Type {
...
   //
-  // For example, consider the deinflection rule that allows 食べ (imperative)
-  // to be de-inflected to 食べる: べ→べる. In this case, the to-type is an
-  // ichidan/ru-verb, while the from type is basically anything but NOT the
-  // result of any other deinflection (since they never produce verb stems).
-  // For this case we set the from-type to Initial so that it does NOT match
-  // any of the existing word types but it DOES match when we compare with 0xff
-  // (the type we use for the initial input).
+  // Similarly, the type Intial ensures that a rule is only applied when it is
+  // the last part of an inflecting word, as every to type is different from it.
```
Comment (birtles) [inline src/background/deinflect.ts:117]: Is is the last part? I thought it was the first part?

File: src/background/deinflect.ts [hunk 4]
```diff
@@ -951,7 +124,294 @@ interface DeinflectRule {
   fromType: number;
   toType: number;
   reason: Reason;
-}
+};
+
+// prettier-ignore
+const deinflectRuleData: Array<DeinflectRule> = [
```
Comment (birtles) [inline src/background/deinflect.ts:130]: Sorry! I meant that we should just change the typing here to be a named tuple, not that we should change it to use `DeinflectRule`.

i.e.

```diff
-const deinflectRuleData: Array<[string, string, number, number, Reason]> = [
+const deinflectRuleData: Array<
+  [from: string, to: string, fromType: number, toType: number, reason: Reason]
+> = [
```

File: src/background/deinflect.ts [hunk 5]
```diff
@@ -1057,7 +516,7 @@ export function deinflect(word: string): CandidateWord[] {
 
         const newWord =
           word.substring(0, word.length - rule.from.length) + rule.to;
-        if (newWord.length <= 1) {
+        if (!newWord.length) {
```
Comment (birtles) [inline src/background/deinflect.ts:519]: What's the thinking behind this change? We now need to deinflect single character strings? Is that correct?

**Comment by birtles** [comment]
> Update:
> 
>     * To be honest, I struggled a bit being precise and concise updating the `DeinflectRule` description. I hope it is clear enough now. Otherwise I'm very open for any suggestions.

Looks good! I added a few minor suggestions just now.

>     * When changing `DeinflectRule` to a named tuple, we would lose the ability to access the elements by dot notation later, like in `getDeinflectRuleGroups()`. When changing DeinflectRule to an object the whole `deinflectRuleData` becomes very verbose. What do you think is the best approach?

Sorry, I didn't mean to change `DeinflectRule` to a named tuple, just the typing for `deinflectRuleData`!

> I will squash the commits after your review.

Thank you! Don't worry if it's too much work, but if it's possible to have one commit where we move the order of the `deinflectRuleData` and related types and one commit where we make the other changes I think that would produce the easiest commit history to bisect.

**Comment by birtles** [comment]
> Thinking about it, I’d like to change the "reason" field to an array to eliminate the need for compound forms like PolitePast, PolitePastNegative, etc. This would be an user-facing change because for example, 食べませんでした would be parsed as `< polite < past < negative` (or better `< polite < negative < past` as this reflects the grammatical structure more closely) instead of `< polite past negative`.

I think "polite past negative" might be preferable to most Japanese learners. Even if "< polite < negative < past" is more accurate linguistically, I think most learners learn ～ませんでした as a set and are used to thinking of it as such so I think "< polite < negative < past" might actually be less useful.

> In the future, this would allow us to parse forms like 食べへんかった as `< dialectal < negative < past` or 食べまして as `< polite < te`.

I've been looking forward to tackling dialects for a while. This will be great! I think in this example I'd actually prefer the annotation for 食べへんかった to read "(Kansai-ben) < negative < past".

> Would you be okay with this change? And if so, should that be a separate PR?

Given that this is a functional change, I think a separate PR / issue would be better so we can discuss the user-facing impact.

Thank you so much!

File: src/background/deinflect.ts (hunk 3 above)
Comment (Enellis) [reply src/background/deinflect.ts:117]: Well, depending on the perspective, I guess. It's the first part getting parsed but the last part structurally, i.e. the end of the word.

File: src/background/deinflect.ts (hunk 5 above)
Comment (Enellis) [reply src/background/deinflect.ts:519]: Yes, because the masu-stem of くる and する being single characters.

File: src/background/deinflect.ts (hunk 4 above)
Comment (Enellis) [reply src/background/deinflect.ts:130]: Ah, my bad. This change makes a lot more sense to me now :)

File: src/background/deinflect.ts (hunk 3 above)
Comment (Enellis) [reply src/background/deinflect.ts:117]: I will clarify what is meant.

File: src/background/deinflect.ts [hunk 6]
```diff
@@ -113,8 +113,9 @@ type DeinflectRule = {
   // match it. "食べろます" would be parsed as "< imperative < polite", which
   // is obviously not correct.
   //
-  // Similarly, the type Intial ensures that a rule is only applied when it is
-  // the last part of an inflecting word, as every to type is different from it.
+  // Similarly, the type Intial ensures that a rule is only applied when it
+  // is at the very end of an inflecting word, as every to-type is different
+  // from it.
```
Comment (Enellis) [inline src/background/deinflect.ts:118]: @birtles better?

Comment (birtles) [reply src/background/deinflect.ts:118]: Sounds good!

**Comment by Enellis** [comment]
> Don't worry if it's too much work, but if it's possible to have one commit where we move the order of the `deinflectRuleData` and related types and one commit where we make the other changes I think that would produce the easiest commit history to bisect.

The changes inside the `deinflectRuleData` array kind of depend on the changes around it, so i went with
- Moving `DeinflectRule` and `Type` before `deinflectRuleData` and condense the array elements into a single line.
- Everything else.

Is this acceptable as well?

**Comment by Enellis** [comment]
> I think "polite past negative" might be preferable to most Japanese learners. Even if "< polite < negative < past" is more accurate linguistically, I think most learners learn ～ませんでした as a set and are used to thinking of it as such so I think "< polite < negative < past" might actually be less useful.

_<sup><sub>(I really wonder why it became the standard learning Japanese in a complicated way. ^^)</sub></sup>_
Yeah, I can see that. Removing the compound forms was just a quick idea which isn't really necessary. The switch to an array would have some other good advantages.


> I've been looking forward to tackling dialects for a while. This will be great! I think in this example I'd actually prefer the annotation for 食べへんかった to read "(Kansai-ben) < negative < past".
> [...]
> Given that this is a functional change, I think a separate PR / issue would be better so we can discuss the user-facing impact.

Cool, I have a working branch for better 敬語 support, which would also benefit from the possibility to add multiple reasons to one rule (ませ as `< polite < imperative`). I think I want to focus on that first and then open a new issue and take care of some 方言・関西弁.

**Comment by birtles** [comment]
Sorry for the delay in reviewing. I wanted to go through it properly on a large screen. It looks great. Thanks for splitting up the commits too—that made it much easier to review.

**Comment by birtles** [comment]
> > I think "polite past negative" might be preferable to most Japanese learners. Even if "< polite < negative < past" is more accurate linguistically, I think most learners learn ～ませんでした as a set and are used to thinking of it as such so I think "< polite < negative < past" might actually be less useful.
> 
> _(I really wonder why it became the standard learning Japanese in a complicated way. ^^)_ Yeah, I can see that. Removing the compound forms was just a quick idea which isn't really necessary. The switch to an array would have some other good advantages.

I'm not too worried which form we end up with. My initial reaction to "< polite < negative < past" was just that it looked complicated for something that is reasonably common but maybe that's more to do with the "<" delimiters. Maybe we should consider another way of combining the reasons?

> > I've been looking forward to tackling dialects for a while. This will be great! I think in this example I'd actually prefer the annotation for 食べへんかった to read "(Kansai-ben) < negative < past".
> > [...]
> > Given that this is a functional change, I think a separate PR / issue would be better so we can discuss the user-facing impact.
> 
> Cool, I have a working branch for better 敬語 support, which would also benefit from the possibility to add multiple reasons to one rule (ませ as `< polite < imperative`).

Sounds great. I have long-term ambitions of doing some really helpful things with お聞きしたい, ご利用になる, 御手紙, おビール etc. but the implementation for that will be quite different.

> I think I want to focus on that first and then open a new issue and take care of some 方言・関西弁.

By all means. Looking forward to it! Thanks again!

**Comment by Enellis** [comment]
> Sorry for the delay in reviewing. I wanted to go through it properly on a large screen. It looks great. Thanks for splitting up the commits too—that made it much easier to review.

Absolutely no worries. Thank you for reviewing and merging this!

> I'm not too worried which form we end up with. My initial reaction to "< polite < negative < past" was just that it looked complicated for something that is reasonably common but maybe that's more to do with the "<" delimiters. Maybe we should consider another way of combining the reasons?

Maybe it would help already making them a little bit smaller, like "‹ polite ‹ negative ‹ past"? (I will open a thread in Discussions for this.)

> Sounds great. I have long-term ambitions of doing some really helpful things with お聞きしたい, ご利用になる, 御手紙, おビール etc. but the implementation for that will be quite different.

Interesting. I can imagine this to be a little bit more complicated to implement, but it would be a nice level up for 10ten.

**Pull request merged**